database/src/ops/block.rs

```diff
@@ -1,89 +1,398 @@
 //! Blocks.
 
 //---------------------------------------------------------------------------------------------------- Import
+use crate::{
+    database::{DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    tables::{BlockHeights, BlockInfoV1s, BlockInfoV2s, BlockInfoV3s},
+    transaction::{TxRo, TxRw},
+    types::{BlockHash, BlockHeight, BlockInfoLatest, BlockInfoV1, BlockInfoV2, BlockInfoV3},
+};
 
-//---------------------------------------------------------------------------------------------------- Free Functions
+//---------------------------------------------------------------------------------------------------- Free functions
 /// TODO
-pub fn add_block() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_block<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height: BlockHeight,
+    block: &BlockInfoLatest,
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    add_block_v3(env, tx_rw, height, block)
 }
```

Comment:
There's a function per `BlockInfoV*` type and a single `add_block()` that forwards to the latest type, `BlockInfoV3` in this case.

This is the case for `add_block()` as well - is this ok?

Reply:
Which field is this? Something like:

```rust
pub fn add_block<'env, Ro, Rw, Env>(
    env: &Env,
    tx_rw: &mut Rw,
    block: cuprate_consensus::block::VerifiedBlockInformation
) -> Result<(), RuntimeError>
where
    Ro: TxRo<'env>,
    Rw: TxRw<'env>,
    Env: EnvInner<'env, Ro, Rw>,
{
    match block.block.header.major_version {
        0..v1 => add_to_v1_table(block),
        v2..v3 => add_to_v2_table(block),
        v3 => add_to_v3_table(block),
        _ => ???
    }
}
```

Reply:
yp but v3 doesn't mean hf 3 btw:

v1 = up to HF 4  
v2 = HF 4 to 10
v3 = HF 10 onwards